GoogleForCreators/web-stories-wp
# Design System: Show context menu in a portal

## Context

~The context menu rendered an invisible mask that covered the entire screen (behind the menu). When a user clicked on this mask, the menu closed. However, to interact with the app they would have to click the app again. This wasn't great UX.~

Turns out this is how browsers usually do it. We moved the menu to be in a portal so that clicking anywhere on the screen would close it.

## Summary

1. ~Context menu now calls `onDismiss` when it loses focus. The `Mask` component has been removed and users are able to interact with the app even if the context menu is open in the dashboard and editor.~ The context menu is now displayed through a portal. Clicking anywhere on the screen will close it. Right clicking the mask will also close it.
2. Combined the `AnimatedContextMenu` and `ContextMenu` into one component. The animation can be used by using the `animate` prop.

## Relevant Technical Choices

n/a

## To-do

n/a

## User-facing changes



## Testing Instructions

New Functionality:
- Clicking away from the menu should close it but also interact with the thing you're clicking on. (ex: button or another element)

Everything else should work the same.

Dashboard:
- Should be able to open the menu by clicking or with the keyboard in the `My Stories` view
- Should be able to navigate the menu with keyboard.
- Tabbing away from the menu should close it.
- Arrow keys should not close the menu.
- When navigating with keyboard you should see the focus border.
- All actions in the menu should work as before.

Editor:
- Should be able to open the context menu by 'right-clicking' anywhere in the canvas. This should select the element that is under the right click.
- Right clicking another element should close the previous menu and open a new menu. The clicked element should now be selected
- All actions in the menu should work as before.

### QA




QA should only happen in the dashboard. The right-click menu in the editor will be QA'd all at once

## Reviews

### Does this PR have a security-related impact?

no

### Does this PR change what data or activity we track or use?

no

### Does this PR have a legal-related impact?

no

## Checklist




---



Fixes #8702
